kafka-client/src/main/java/dev/responsive/kafka/clients/OffsetRecorder.java

```diff
@@ -0,0 +1,163 @@
+package dev.responsive.kafka.clients;
+
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.TopicPartition;
+
+/**
+ * Record offsets written to kafka and committed to consumer groups. When a commit happens,
+ * notify listeners about the set of offsets.
+ *
+ * The callbacks here need to be synchronized. Almost all the callbacks happen on the stream
+ * thread. The lone exception is the Send callback, which happens on Producer I/O threads.
+ */
+class OffsetRecorder {
+  private final Map<RecordingKey, Long> uncommitted = new HashMap<>();
+  private final Map<TopicPartition, Long> written = new HashMap<>();
+  private final ProducerListener producerListener = new ProducerListener();
+  private final ConsumerListener consumerListener = new ConsumerListener();
+  private final List<CommitCallback> commitCallback = new LinkedList<>();
+  private final boolean eos;
+
+  OffsetRecorder(final boolean eos) {
+    this.eos = eos;
+  }
+
+  public synchronized void addCommitCallback(final CommitCallback callback) {
+    commitCallback.add(callback);
+  }
+
+  public synchronized void recordConsumedOffsets(
+      final String consumerGroup,
+      final TopicPartition partition,
+      final long offset
+  ) {
+    uncommitted.put(new RecordingKey(partition, consumerGroup), offset);
+  }
+
+  public synchronized void recordProduce(final RecordMetadata recordMetadata) {
+    final TopicPartition partition
+        = new TopicPartition(recordMetadata.topic(), recordMetadata.partition());
+    written.compute(
+        partition,
+        (k, v) -> v == null ? recordMetadata.offset() : Math.max(recordMetadata.offset(), v)
+    );
+  }
+
+  public synchronized void recordCommit() {
+    final Map<RecordingKey, Long> committedOffsets = Map.copyOf(uncommitted);
+    final Map<TopicPartition, Long> writtenOffsets = Map.copyOf(written);
+    commitCallback.forEach(c -> c.onCommit(committedOffsets, writtenOffsets));
```

Comment:
these callbacks will all run within the synchronized block, I suspect that was intentional? If I'm following the code correctly that means we're flushing to the remote within the synchronized block, is there any reason that's necessary? We're making copies of things anyway, and if (again) I'm reading the code correctly that means each thread will block on one another if commits happen concurrently.

on that thought, why does this need to be synchronized at all? It looks like it's only created once per stream thread?

Reply:
yeah I suppose we don't need to do the actual flushes when synchronized (though it's not the end of the world as this should be the only thread trying to acquire this lock at that point). The reason its synchronized is because the producer send callbacks happen on the producer I/O threads.